rust/src/livefs.rs

```diff
@@ -0,0 +1,367 @@
+/*
+ * Copyright (C) 2020 Red Hat, Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ */
+
+use anyhow::{bail, Context, Result};
+use nix::sys::statvfs;
+use openat_ext::OpenatDirExt;
+use ostree::DeploymentUnlockedState;
+use serde_derive::{Deserialize, Serialize};
+use std::borrow::Cow;
+use std::os::unix::io::AsRawFd;
+use std::path::{Path, PathBuf};
+use std::process::Command;
+
+const OSTREE_RUNSTATE_DIR: &str = "/run/ostree/deployment-state";
+const LIVEFS_STATE_NAME: &str = "rpmostree-livefs-state";
+
+#[derive(Debug, Default, Clone, Serialize, Deserialize)]
+struct LiveFsState {
```

Comment:
When I was touching this data last time, I was unsure about the actual semantics of this, so let me ask it here. Are `commit` and `inprogress_commit` actually related in some way?
That is, are they maybe in an XOR state (and thus an `enum` here) or perhaps in a linked state (and thus a `Option<LiveFsState>` with non-null fields)?

Either way, it would be good to document what these two fields hold and what is the scenario they represent, to ease the life of casual readers.

Reply:
It's 
```
diff --git a/rust/src/livefs.rs b/rust/src/livefs.rs
index 2dc2bef8..03f9f1a4 100644
--- a/rust/src/livefs.rs
+++ b/rust/src/livefs.rs
@@ -19,7 +19,12 @@ const LIVEFS_STATE_NAME: &str = "rpmostree-livefs-state";
 
 #[derive(Debug, Default, Clone, Serialize, Deserialize)]
 struct LiveFsState {
+    /// The OSTree commit that the running root filesystem is using,
+    /// as distinct from the one it was booted with.
     commit: Option<String>,
+    /// Set when a livefs operation is in progress; if the process
+    /// is interrupted, some files from this commit may exist
+    /// on disk but in an incomplete state.
     inprogress_commit: Option<String>,
 }
 
```

(But not pushing yet since I don't want to break the green CI)

Reply:
Definitely not a XOR - we support going from "base commit" :arrow_right:  livefs(A) :arrow_right:  livefs(B) and if the A :arrow_right: B transition is interrupted, then we have 
```
state {
  commit: Some(A),
  inprogress_commit: Some(B)
}
```

Whereas if B completes then we have
```
state {
  commit: Some(B)
  inprogress_commit: None
}
```

right?

Reply:
Thanks for the insight! No need to amend the PR right noe, it's already good to have these details noted down here.
Aside, in the end it looks like:
 1. this models a two-states `InProgress`/`Active` FSM
 2. in both states the `commit` field is present (is there an actual case where it's `None`?)